### DISCUSSION TITLE
Difference between CONFIG_HW_STACK_PROTECTION and CONFIG_STACK_SENTINEL

### DISCUSSION THREAD

**Discussion by XavierChapron** [zephyrproject-rtos/zephyr#28866] [Q&A] [answered]
Hi,

I'm wondering what should be used between CONFIG_HW_STACK_PROTECTION and CONFIG_STACK_SENTINEL at least regarding NRF MCU.

Does both of them catch every stack overflow?
Is there any reason not to use the hardware version? Maybe in term of footprint, mcu usage, mcu consumption?

Thanks!

Xavier

**Comment by ioannisg** [comment, marked as answer]
The HW stack protection relies on HW-based mechanisms to detect thread stack overflows. Examples:
-  In ARMv8-M, HW_STACK_PROTECTION makes use of the Cortex-M stack pointer limit registers, which trigger a Usage fault if stack pointers descent below their values. 
- In ARMv7-M, HW_STACK_PROTECTION is implemented with a read-only MPU region guard, so a MemManage fault is triggered when a thread attempts to push into the guard area.

Stack sentinel is supposed to be used when the CPU does not have a native HW stack protection mechanism. It marks a "guard" area at the end of the thread stack with a given pattern. Each time a thread switches-out, -in or during IRQ exit, the area is checked whether it still holds the pattern and an CPU error is triggered if not.

HW-based mechanisms mush be used when they are available.
